rusoto/core/src/reactor.rs

```diff
@@ -0,0 +1,225 @@
+            panic!("failed to send request to reactor: {}", err);
+        }
+        RequestDispatcherFuture {
+            receiver: rx
+        }
+    }
+}
+
+/// A credentials provider backed by an implicit event loop.
+pub struct CredentialsProvider {
+    sender: mpsc::UnboundedSender<((), oneshot::Sender<Result<AwsCredentials, CredentialsError>>)>
+}
+
+impl Default for CredentialsProvider {
+    fn default() -> CredentialsProvider {
+        DEFAULT_REACTOR.default_credentials_provider()
+            .expect("failed to create default credentials provider")
+    }
+}
+
+/// Future returned from `CredentialsProvider`.
+pub struct CredentialsProviderFuture {
+    receiver: oneshot::Receiver<Result<AwsCredentials, CredentialsError>>
+}
+
+impl Future for CredentialsProviderFuture {
+    type Item = AwsCredentials;
+    type Error = CredentialsError;
+
+    fn poll(&mut self) -> Poll<Self::Item, Self::Error> {
+        match self.receiver.poll().expect("failed to retrieve response from reactor") {
+            Async::NotReady => Ok(Async::NotReady),
+            Async::Ready(result) => result.map(Async::Ready)
+        }
+    }
+}
+
+impl ProvideAwsCredentials for CredentialsProvider {
+    type Future = CredentialsProviderFuture;
+
+    fn credentials(&self) -> Self::Future {
+        let (tx, rx) = oneshot::channel();
+        if let Some(err) = self.sender.unbounded_send(((), tx)).err() {
+            panic!("failed to send request to reactor: {}", err);
+        }
+        CredentialsProviderFuture {
+            receiver: rx
+        }
+    }
+}
+
+impl Reactor {
+    pub fn default_request_dispatcher(&self) -> Result<RequestDispatcher, TlsError> {
+        self.new_request_dispatcher(|handle| default_tls_client(&handle))
+    }
+
+    pub fn default_credentials_provider(&self) -> Result<CredentialsProvider, CredentialsError> {
+        self.new_credentials_provider(|handle| DefaultCredentialsProvider::new(&handle))
+    }
+
+    pub fn new_request_dispatcher<D: DispatchSignedRequest + 'static, E: Send + 'static, F: FnOnce(Handle) -> Result<D, E> + Send + 'static>(&self, make_dispatcher: F) -> Result<RequestDispatcher, E> {
+        self.new_responder(|handle| {
+            make_dispatcher(handle).map(|dispatcher| move |request| { dispatcher.dispatch(request) })
+        }).map(|sender| RequestDispatcher { sender: sender })
+    }
+
+    pub fn new_credentials_provider<P: ProvideAwsCredentials + 'static, E: Send + 'static, F: FnOnce(Handle) -> Result<P, E> + Send + 'static>(&self, make_provider: F) -> Result<CredentialsProvider, E> {
+        self.new_responder(|handle| {
+            make_provider(handle).map(|provider| move |()| { provider.credentials() })
+        }).map(|sender| CredentialsProvider { sender: sender })
+    }
+
+    fn new_responder<T, U, E, F, G>(&self, make_responder: F) -> Result<mpsc::UnboundedSender<(T, oneshot::Sender<Result<U::Item, U::Error>>)>, E>
```

Comment:
This function is pretty dense to someone without much Futures/Tokio exposure, such as myself.  Can we get some docs as to the purpose of this chunk of code?

Reply:
Will do!